### PR TITLE
feat: add .testgroundignore file

### LINKED ISSUE #1170: Add a .testgroundignore kind-of feature
## Description

We should add a `.testgroundignore` kind of feature (similar to `.gitignore`). I came across this issue while building the `docker/node` builder (#1158). For Node.js projects, the `node_modules` directory is usually quite big, and uploading it to the daemon every time we need to run a test doesn't seem to be the wisest option.

Thus, I suggest creating an ignore list, in the format of a `.testgroundignore` file.

## What defines this endeavour to be complete?

- [ ] Support `.testgroundignore`

### PR DESCRIPTION
**Pull request by laurentsenta**
Fixes #1170

### CODE REVIEW

**Comment by laurentsenta** [comment]
@hacdias 👋  just tagging you if you want to follow along, but no input is required from you.

File: pkg/client/client.go [hunk 1]
```diff
@@ -221,6 +235,45 @@ func (c *Client) runBuild(ctx context.Context, r interface{}, path, plandir, sdk
...
+// getFilteredDirectory filters the directory dir according to the
+// ignored files specified in $dir/.testgroundignore. Returns a new
+// directory and a boolean indicating if such directory is temporary
+// and should be cleaned by the caller.
+func getFilteredDirectory(dir string) (string, bool, error) {
+	ignoreFilePath := filepath.Join(dir, ".testgroundignore")
+	_, err := os.Stat(ignoreFilePath)
+
+	if os.IsNotExist(err) {
+		return dir, false, nil
```
Comment (galargh) [inline pkg/client/client.go:247]: Don't we want to create a temp dir either way?

Is it possible that the dir files get modified during the run? If so, then if we always create a temp dir, no matter if `.testgroundignore` is provided or not, the behaviour is consistent and the file modifications are not reflected after the run.

File: pkg/client/client.go [hunk 2]
```diff
@@ -221,6 +235,45 @@ func (c *Client) runBuild(ctx context.Context, r interface{}, path, plandir, sdk
...
+	} else if err != nil {
+		return "", false, err
+	}
+
+	tgIgnore, err := ignore.CompileIgnoreFile(ignoreFilePath)
```
Comment (galargh) [inline pkg/client/client.go:252]: Could we have a comment that explains that the output of `CompileIgnoreFile` is a set of regex patterns (as per https://github.com/sabhiram/go-gitignore/blob/master/ignore.go#L164)?

It wasn't immediately obvious to me and without this information it's hard to judge if the comparisons we're doing later are correct.

File: pkg/client/client.go (hunk 1 above)
Comment (galargh) [reply pkg/client/client.go:247]: We also wouldn't need the temp boolean anymore if we always create a temp dir.

File: pkg/client/fixtures/with_ignore/.testgroundignore [hunk 3]
```diff
@@ -0,0 +1,3 @@
+a
+b/c
```
Comment (galargh) [inline pkg/client/fixtures/with_ignore/.testgroundignore:2]:
```suggestion
c
```

I think it'd be interesting to test if the above works. If we just have `b/c` here, then a naive implementation that doesn't turn lines here into regexes would still work but if we have only `c` then it wouldn't anymore. Does it make sense?

**Review by galargh — COMMENTED** [review]
Cool change :) Left a few comments here and there but I like it a lot :)

File: pkg/client/client.go (hunk 2 above)
Comment (brdji) [reply pkg/client/client.go:252]: I agree with @galargh  - I would like to see a more comments detailing this function.

Comment (laurentsenta) [reply pkg/client/client.go:252]: Updated

File: pkg/client/client.go [hunk 4]
```diff
@@ -18,6 +18,9 @@ import (
 	"strings"
 	"sync"
 
+	copy2 "github.com/otiai10/copy"
```
[the remark this answers is not in the document]
Comment (laurentsenta) [reply pkg/client/client.go]: Probably a left over, I removed it

File: pkg/client/client.go (hunk 1 above)
Comment (laurentsenta) [reply pkg/client/client.go]: Sounds good, I liked that first approach, but dropping it make sense, fixed.

File: pkg/client/fixtures/with_ignore/.testgroundignore (hunk 3 above)
Comment (laurentsenta) [reply pkg/client/fixtures/with_ignore/.testgroundignore]: Agreed, I updated the test!

File: pkg/client/client.go (hunk 2 above)
Comment (laurentsenta) [reply pkg/client/client.go:261]: Added a comment with some clarification.

File: .github/workflows/integration-tests.yml [hunk 5]
```diff
@@ -37,4 +36,4 @@ jobs:
           make install
       - name: Run tests
         run: |
-          go test -tags "integration,${{ matrix.suite }}" -timeout 20m -parallel 1 -v -shuffle=on ./pkg/integration/...
```
Comment (galargh) [inline .github/workflows/integration-tests.yml:40]: Why this change? At face value, I'd read it as an indication of the test suite getting worse.

**Review by galargh — APPROVED** [review]

File: .github/workflows/integration-tests.yml (hunk 5 above)
Comment (laurentsenta) [reply .github/workflows/integration-tests.yml:40]: Agreed, but we're using 1.16 for now, and that option was added with go 1.17, it's messing up with the tests,
I added a comment to clarify this.
